Dev

vlad-marlo opened the pull request.

vlad-marlo committed:
migrated app from old repo

vlad-marlo committed:
migrated app from old repo

vlad-marlo committed:
Merge remote-tracking branch 'origin/dev' into dev

vlad-marlo committed:
added new migrations on db

vlad-marlo committed:
wrote default db migrations;

vlad-marlo committed:
fixed tests

vlad-marlo committed:
wrote more tests

vlad-marlo committed:
created roles repository;
configured swagger docs;
minor changes in configuration;
just do it

vlad-marlo committed:
created auth tokens
more swagger docs

vlad-marlo committed:
wrote more tests and functionality
minor changes in database (now invite include role in group in it)

vlad-marlo committed:
Merge pull request #1 from vlad-marlo/tests

Tests

vlad-marlo committed:
added docker support to project

vlad-marlo committed:
minor changes;
added some docstrings;
decomposed addition and getting user from request in mw;
created addition task to group in repository;

vlad-marlo committed:
minor changes;
added some docstrings;
decomposed addition and getting user from request in mw;
created addition task to group in repository;

vlad-marlo committed:
idk how to comment this

vlad-marlo committed:
changed dependence from user id in string to uuid.UUID in service;
fixed tests;

vlad-marlo committed:
created dockerfiles;
created invite handlers - create, use;
some minor changes in server structure

vlad-marlo committed:
some tests on storage

vlad-marlo committed:
created register and create token grpc handlers

vlad-marlo committed:
some fucking changes idk what changed maybe new Me endpoint that return user profile - groups, tasks etc.

vlad-marlo committed:
minor changes;
added 2 new endpoints - get all user's tasks and get task by id;

vlad-marlo committed:
my head is spinning like a screw;
added create task handler;
fucking sql I hate it

vlad-marlo committed:
Merge branch 'dev' into invites

vlad-marlo committed:
Merge pull request #2 from vlad-marlo/invites

Invites

vlad-marlo committed:
wrote some tests for http controller(now 46.6 % cov)

vlad-marlo committed:
wrote some tests for http controller(now 46.6 % cov)

tmvrus commented on cmd/server/main.go:
```diff
@@ -0,0 +1,137 @@
+			pgx.NewInviteRepository,
+			httpctrl.New,
+		),
+		//fx.WithLogger(ZapEventLogger),
```
Такие комментарии засоряют рабочий код и усложняют его восприятие

tmvrus commented on cmd/server/main.go:
```diff
@@ -0,0 +1,137 @@
+
+// _ return new event logger for fx application.
+func _(logger *zap.Logger) fxevent.Logger {
+	return &fxevent.ZapLogger{Logger: logger}
```
эта ф-я где-то используется?
выглядит как неудачное переименование

tmvrus commented on cmd/server/main.go:
```diff
@@ -0,0 +1,137 @@
+func ValidateConfig(cfg *config.Config, log *zap.Logger) error {
+	if ok, err := cfg.Valid(); err != nil || !ok {
+		log.Error("config is not valid", zap.Bool("ok", ok), zap.Error(err))
+		return errors.New("bad config")
```
тут хорошо бы сообщить клиенту что именно не так с конфигом

tmvrus commented on internal/config/config.go:
```diff
@@ -0,0 +1,289 @@
+		c.Server.EnableHTTP = true
+	case "GRPC", "grpc":
+		c.Server.EnableGRPC = true
+	}
```
тут можно было сделать switch strings.ToLower(c.Server.Type)
и сократить количество вариантов в два раза

tmvrus commented on internal/config/config.go:
```diff
@@ -0,0 +1,289 @@
+	}
+	ok, err = govalidator.ValidateStruct(c)
+	// validate that enabled
+	ok = ok && (c.Server.EnableGRPC != c.Server.EnableHTTP) && (c.Server.EnableHTTP || c.Server.EnableGRPC) && strings.Count("%s", c.Server.InviteLinkTemplate) != 3
```
плохо читается условие, такое лучше разбивать на несколько строк

tmvrus commented on internal/config/config.go:
```diff
@@ -0,0 +1,289 @@
+}
+
+func (c *Config) WriteToFile(file string) error {
+	_ = os.Remove(file)
```
если нужно сделать опциональный rewrite файла, то можно передать os.O_CREATE|os.O_TRUNC при открытии и ничего не нужно будет удалять

tmvrus commented on internal/controller/grpc/handlers.go:
```diff
@@ -0,0 +1,60 @@
+func (s *Server) Ping(ctx context.Context, _ *pb.PingRequest) (*pb.PingResponse, error) {
+	var resp pb.PingResponse
+	if err := s.srv.Ping(ctx); err != nil {
+		fieldErr, ok := err.(*fielderr.Error)
```
вот эта логика много раз встречается в проекте, это признак того что ее лучше убрать у ф-ю типа fielderr.Is(err error)

внутри лучше использовать не type assert `err.(*fielderr.Error)` а errors.As, это концептуально правильней

tmvrus commented on internal/controller/grpc/server.go:
```diff
@@ -0,0 +1,77 @@
+//go:generate mockgen --source=server.go --destination=mocks/service.go --package=mocks
```
код генерации лучше писать после импорта, а это место оставить для документации к пакету, иногда создают отдельный файл generate.go и собирают всю генерацию там

tmvrus commented on internal/controller/grpc/server.go:
```diff
@@ -0,0 +1,77 @@
+	return nil
+}
+
+func (s *Server) Stop(context.Context) error {
```
контекст передается не используется (

vlad-marlo replied:
это необходимо для создания хука в zap.fx

tmvrus commented on internal/controller/http/handlers.go:
```diff
@@ -0,0 +1,494 @@
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
```
судя по документации - клиенту нужно отдавить OK в этом месте

tmvrus commented on internal/controller/http/handlers.go:
```diff
@@ -0,0 +1,494 @@
+
+	var req model.CreateGroupRequest
+	if err := json.NewDecoder(&buf).Decode(&req); err != nil {
+		s.respond(w, http.StatusBadRequest, nil, ReqIDField(reqID), zap.Error(err))
```
не хватает return

tmvrus commented on internal/controller/http/handlers.go:
```diff
@@ -0,0 +1,494 @@
+	_ = r.Body.Close()
+
+	if err := json.NewDecoder(&buf).Decode(&req); err != nil {
+		s.respond(w, http.StatusInternalServerError, nil, zap.Error(err), ReqIDField(reqID))
```
return

tmvrus commented on internal/controller/http/handlers.go:
```diff
@@ -0,0 +1,494 @@
+	}
+
+	if err = json.NewDecoder(&buf).Decode(&req); err != nil {
+		s.respond(w, http.StatusInternalServerError, nil, zap.Error(err), ReqIDField(reqID))
```
return
и если мы сервис не может распасить json то это скорее BadRequest

tmvrus commented on internal/controller/http/handlers.go:
```diff
@@ -0,0 +1,494 @@
+
+	var req model.TaskCreateRequest
+	if err := json.NewDecoder(&buf).Decode(&req); err != nil {
+		s.respond(w, http.StatusBadRequest, nil, zap.Error(err), reqID)
```
return

tmvrus commented on internal/controller/http/middleware/auth.go:
```diff
@@ -0,0 +1,89 @@
+func respond(w http.ResponseWriter, code int, data interface{}, fields ...zap.Field) {
+	var lvl zapcore.Level
+	switch {
+	case code >= 500:
+		lvl = zap.DPanicLevel
+	case code >= 400:
+		lvl = zap.ErrorLevel
```
вместо цифр лучше использовать константы http.StatusXXX так нагляднее

vlad-marlo committed:
wrote some tests:
100% cov of fielderr pkg;
most of http handlers covered >80%;
some minor tests;

vlad-marlo committed:
57.8% coverage
has to do do do do do do do do do do do

vlad-marlo committed:
wrote some test => now 60.4% cov

tmvrus commented on internal/service/interface.go:
```diff
@@ -0,0 +1,33 @@
+	GetUserTasks(ctx context.Context, user uuid.UUID) (*model.GetTasksResponse, error)
+	// GetTask return task by id if user is related to it.
+	GetTask(ctx context.Context, user, task uuid.UUID) (*model.Task, error)
+	// CreateTask ...
```
такие комментарии лишены смысла, возможно, стоит вообще от них отказаться

tmvrus commented on migrations/004_tasks.sql:
```diff
@@ -0,0 +1,35 @@
+    description text,
+    created_at  timestamp default current_timestamp,
+    created_by  uuid not null,
+    status text not null default 'NEW',
```
поля такого типа лучше описывать в вид enum - это более "error prone" и гораздо читабельнее: сразу понятно каки значения есть у поля

tmvrus commented on internal/store/pgx/invites.go:
```diff
@@ -0,0 +1,154 @@
+		&group,
+		&role,
+	); err != nil {
+		if pgErr, ok := err.(*pgconn.PgError); ok {
```
это условие встречается в нескольких местах - хороший кандидат на helper-функцию

tmvrus commented on internal/service/production/tasks.go:
```diff
@@ -0,0 +1,88 @@
+		s.log.Warn("error while getting role of member", zap.Error(err))
+		return
+	}
+	if r.Tasks >= 2 {
```
классическая magic number, смысл цифры 2 тут ясен только автору кода, что бы это исправить лучше использовать константы
https://en.wikipedia.org/wiki/Magic_number_(programming)#Unnamed_numerical_constants

tmvrus commented on internal/service/production/tasks.go:
```diff
@@ -0,0 +1,88 @@
+		return nil, service.ErrInternal.With(zap.Error(err))
+	}
+
+	// async add task to group and users
+	go s.addTaskToGroup(ctx, user, task.ID, req.Group)
+	go s.addToUsers(ctx, user, task.ID, req.Users)
+
```
в эти горутины лучше не передавать контекста запроса, потому что подразумевается их независимое выполнение, какой-нибудь код снаружи вполне может этот контекст закрыть после окончания обработки запроса, а это повлечет за собой ошибки в горутинах

tmvrus commented on internal/store/pgx/invites.go:
```diff
@@ -0,0 +1,154 @@
+		return Unknown(err)
+	}
+	defer func() {
+		_ = tx.Rollback(ctx)
```
ошибки лучше все же логировать

tmvrus commented on internal/service/production/groups_test.go:
```diff
@@ -0,0 +1,125 @@
+package production
```
хорошей практикой является вынесение тестов в отдельный пакет `production_test` (при таком именование можно оставить файлы в той же папке) и тестировать код как "black box"

tmvrus reviewed (changes requested):
Что хорошо:
- проект выглядит зрелым
- есть миграции
- есть Makefile
- поддержка docker
- есть документация

Что надо подправить:
- код должен быть покрыт тестами на 80%, как минимум пакеты service и store
- не везде отсортированы импорты, достадная оплошность
- хорошо бы пройтись gofmt по всему проекту
- есть мелкие стилистически замечания
- много пропущенных return в http-layer

На мой взгляд, go.uber.org/fx в проекте лишний. Фактически тут три компонента - api,storage и service, fx все же нужен для стандартизации проектов крупнее.

vlad-marlo committed:
added docstrings;
fixed some style issues;
xD;

vlad-marlo committed:
added docstrings and test;
now +0.5% coverage.

vlad-marlo committed:
go fmt;
added some test;
changed some things in storage;
i like ya practicum<3;

vlad-marlo committed:
now 70.7% coverage

vlad-marlo committed:
fixed some minor fixes

vlad-marlo committed:
Merge pull request #4 from vlad-marlo/tests

wrote some tests for http controller(now 46.6 % cov)

tmvrus commented on internal/config/config.go:
```diff
@@ -0,0 +1,299 @@
+	"go.uber.org/zap"
+)
+
+//goland:noinspection ALL,GoUnnecessarilyExportedIdentifiers,GoUnnecessarilyExportedIdentifiers,GoUnnecessarilyExportedIdentifiers,GoUnnecessarilyExportedIdentifiers,GoUnnecessarilyExportedIdentifiers,GoUnnecessarilyExportedIdentifiers
```
первый раз вижу что бы кто-то пользовался этим

tmvrus reviewed (approved):
У хорошего проекта должен быть хороший readme файл)
